### PR TITLE
ipn/ipnlocal: prevent putting file if file already exists

### PR DESCRIPTION
**Pull request by rhealitycheck**
Also adding tests to ensure this works.

Updates tailscale/corp#14772

### CODE REVIEW

**Review by phirework — APPROVED** [review]

File: ipn/ipnlocal/peerapi.go [hunk 1]
```diff
@@ -1122,6 +1122,13 @@ func (h *peerAPIHandler) handlePeerPut(w http.ResponseWriter, r *http.Request) {
...
 	// TODO(bradfitz): prevent same filename being sent by two peers at once
+
+	// prevent same filename being sent twice
+	if _, err := os.Stat(dstFile); err == nil {
+		http.Error(w, "file exists", http.StatusConflict)
+		return
+	}
+
 	partialFile := dstFile + partialSuffix
 	f, err := os.Create(partialFile)
```
[the remark this answers is not in the document]
Comment (dsnet) [reply ipn/ipnlocal/peerapi.go:1133]: 👍 , let's do this change as well.

**Review by bradfitz — COMMENTED** [review]
Extra slash in commit message bug reference.

Is this for after 1.50 is cut?

**Review by bradfitz — COMMENTED** [review]
What's the goal here? To stop dups at the first hop only or all the way to the final location?

**Comment by dsnet** [comment]
> What's the goal here? To stop dups at the first hop only or all the way to the final location?

Conflicts aren't really an issue in platforms that eagerly move files outside the waiting directory (e.g., Windows, Android, etc.), but is a problem on platforms like Linux where files just sit there for a while. (This can technically still be an issue on Windows and friends in rare race situations).

In general, Taildrop should work hard to never overwrite files, which leads us to conclude that it's better to reject an existing file than to silently overwrite it.

A future PR will expand the logic to have better conflict handling, but we're aiming to do the easy, but safe thing now.

**Comment by dsnet** [comment]
> Is this for after 1.50 is cut?

This isn't pressing, so it could wait till after the cut.

**Comment by rhealitycheck** [comment]
> Extra slash in commit message bug reference.
> 
> Is this for after 1.50 is cut?

Yeah I was expecting it to be for after the 1.50 cut.

File: ipn/ipnlocal/peerapi.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rhealitycheck) [reply ipn/ipnlocal/peerapi.go:1133]: So I discussed with @dsnet and @andrew-d separately but if we do this change then it means that any partial file we try to write that already exists would be blocked which is problematic in the case where a file fails to send and there's a partial leftover but we want to retry to get the whole file over. I think we can address this but as a separate fix.

**Pull request merged**